File: 347.top_k_frequent_elements/topKFrequent.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function(nums, k) {
    var frequenceMap = new Map();
    for(var elem in nums){//set up the frequence map
        var val = frequenceMap.get(nums[elem]);
        if(val === undefined){
            frequenceMap.set(nums[elem],1);
        }else{
            frequenceMap.set(nums[elem],val+1);
        }
    }
    
    var bucket = new Array(nums.length+1);
    frequenceMap.forEach(function(freq,num){
        if(bucket[freq] === undefined){
            bucket[freq] = [];
        }
        bucket[freq].push(num);
    });
    
    var result = [];
    for(var i = bucket.length-1; i>0 && k>0; i--){
        if(bucket[i] !== undefined){
            result = result.concat(bucket[i]);
            k -= bucket[i].length;
        }
    }
    
    return result;
    
    
};

console.log(topKFrequent([1,1,1,2,2,3],2));

//notes:
// for(var elem in nums)  // elem is the index

//array concatenate: result = result.concat(bucket[i]);

File: 347.top_k_frequent_elements/topKFrequent.py
class Solution(object):
    def topKFrequent(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        freq = {}
        for num in nums:
            if freq.has_key(num):
                freq[num] += 1
            else:
                freq[num] = 1

        bucket = [0]*(len(nums)+1)
        for num, freq in freq.items():
            if (bucket[freq] == 0):
                bucket[freq] = []
            bucket[freq].append(num)

        result = []
        i = len(bucket) - 1
        while i > 0 and k > 0:
            if (bucket[i] != 0):
                result += bucket[i]
                k -= len(bucket[i])
            i -= 1
        return result

s = Solution()
print s.topKFrequent([1,1,1,2,2,3], 2)